### PR TITLE
Final-Changes_and_documentations: Changes  

### PR DESCRIPTION
**Pull request by Paul-tes**
### Foolwed Design
- implemented the following [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
Colors.
Typography: font face, size and weight.
Layout: composition and space between elements.
- Original design idea by [Gregoire Vella on Behance](https://www.behance.net/gregoirevella).

#### Splash screen
- Added A simple page with the name of the app.  and links to the sign up and log in pages.

#### Sign up and log in pages
- Added, A user to register in the app with full name, email and password (all mandatory).
- Added, A user can log into the app using email and password.
- Added conditions If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

#### Home page (categories page)

- Add categories page, when the user logs in.
- Added a button, when the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
-Added a button "add a new category" at the bottom that brings the user to the page to create a new category.

#### Purchase page

- Added For a given category, the list of transactions is presented, ordered by the most recent.
-  Added user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
- Added a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
- Added a "Back" button (<), the user navigates to the home page.

- Add a new category" page
   - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
   - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
   - When the user clicks on the "Back" button (<), the user navigates to the home page.

- Add a new transaction" page
   - The user fills out a form to create a new transaction with:
   - name (mandatory)
   - amount (mandatory)
   - categories (mandatory at least one)
   - Added  "Save" button to create the new transaction, and is taken to the transactions page for that category.
   - Added a "Back" button (<), the user navigates to the transactions page for that category.
#### Technical changes
- Used Postgres as your database.
- Used devise for authentication.
- Added validation to validate all user input to make sure that anyone with bad intentions cannot compromise your app.
- Used view template engine .erb

### CODE REVIEW

File: app/views/categories/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    </div>
+  <% end %>
+
+  <div class="field input-cont mt-12 rounded">
+    <%= form.text_field :name, placeholder: :Name, class: "input"%>
+    <%= form.text_field :icon, placeholder: :icon, class: "input"%>
+    <div class="p-2 bg-blue text-center text-white rounded mt-20 w-20 mx-auto">
+      <%= form.submit "Save" %>
+    </div>
+  </div>
```
Comment (muneebulrehman) [inline app/views/categories/_form.html.erb]: When creating a category, there is no submit button visible on the UI. Kindly fix this issue. ⬇️
image

File: app/views/purchases/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  <div class="field input-cont mt-12 rounded">
+    <%= form.text_field :name, placeholder: :Name, class: "input" %>
+    <%= form.text_field :amount, placeholder: :Price, class: "input" %>
+    <div class="flex justify-between items-start my-6">
+      <%= form.collection_select :category_id, current_user.categories, :id, :name, { include_blank: false }, {class: "bg-gray-200 text-gray-700 border border-gray-200 rounded text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-40 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 mx-5"} %>
+    </div>
+    <div class="p-2 bg-blue text-center text-white rounded mt-20 w-20 mx-auto">
+      <%= form.submit "Save" %>
+    </div>
+  </div>
```
Comment (muneebulrehman) [inline app/views/purchases/_form.html.erb]: While creating a new purchase, there is no submit button visible on the UI. Kindly fix this issue. ⬇️
image

File: app/views/categories/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<div id="categories" class="body-height custom-light-gray-bg">
+  <% @categories.each do |category| %>
+    <%= render category %>
+  <% end %>
+  <%= link_to "Add category", new_category_path(user_id: current_user), class: "px-4 py-2 fixed bottom-0 text-center w-full font-semibold text-sm bg-cyan-600 text-white rounded-none shadow-sm hover:bg-cyan-700" %>
```
Comment (muneebulrehman) [inline app/views/categories/index.html.erb]: The**Add Category** button should look similar to the button provided in the original design. Kindly follow-up on the same. It is a project requirement to stick to the provided design.
image

File: app/views/purchases/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+      <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-6 h-6 text-black">
+        <path stroke-linecap="round" stroke-linejoin="round" d="M10.5 19.5L3 12m0 0l7.5-7.5M3 12h18" />
+      </svg>
+    <%end%>
+    <p class="font-black"><%=number_to_currency(@purchases_sum)%></p>
+  </div>
+  <% @purchases.each do |purchase| %>
+    <%= render purchase %>
+  <% end %>
+  <%= link_to "New purchase", new_category_purchase_path, class: "px-4 py-2 fixed bottom-0 text-center w-full font-semibold text-sm bg-cyan-600 text-white rounded-none shadow-sm hover:bg-cyan-700" %>
```
Comment (muneebulrehman) [inline app/views/purchases/index.html.erb]: The **New purchase** button should look similar to the one provided in the original design. Kindly update this component accordingly. ⬇️
image

**Review by muneebulrehman — CHANGES_REQUESTED** [review]
Hi @Paul-tes , 

Good job so far! I see you put a lot of effort into this project. However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- No linter errors
- Descriptive PR
- No linter errors
- App functionality working as expected.

## Required Changes ♻️
_Check the comments under the review._

- [x] As per the project requirements, you should record a presentation video following the below guidelines. However, it is missing in your case. Kindly record the same before submitting for a re-review. ⬇️
image

- [x] Some of the test cases are failing. Kindly fix these before submitting for a re-review.
image



## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Paul-tes** [comment]
> I have noticed the test is working on my machine successfully. please check the right branch on the updated version.
pass_test

**Review by OyePriscilla — APPROVED** [review]
Hi @Paul-tes 👏 👏 👏, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
All test cases passed ✔️ 
Designed UI in accordance to the given template ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**